### ISSUE TITLE
Review current password rules and update them to enforce best practices

### ISSUE THREAD

**Issue by ram-kishore-odi** [cagov/data-infrastructure#442]
The main goal of this ticket is to review the current password rules associated with snowflake accounts and update them to enforce best practices. The same rules will need to be used to setup break glass accounts

**Comment by ram-kishore-odi** [comment]
Following are best practices for password rules in Snowflake, based on their official documentation, to enforce strong security for regular and break-glass accounts - 

1. Password Complexity:

Length: Enforcing a minimum password length is critical. Snowflake supports a range of 8 to 256 characters. A minimum of **14** characters is recommended. 
PASSWORD_MIN_LENGTH = 14

Character Diversity: Our policy wille  a mix of uppercase, lowercase, numbers, and special characters (!@#$%^&* etc.).
PASSWORD_REQUIRE_UPPERCASE = true
PASSWORD_REQUIRE_LOWERCASE = true
PASSWORD_REQUIRE_NUMERIC = true
PASSWORD_REQUIRE_SPECIAL_CHAR = true

2. Password Expiration:

Regular Expiration: Setting an appropriate password expiration interval is also very important. Snowflake's default is 90 days. Our plan is consider using a shorter interval like 60 for enhanced security. If the passwords are older than 60 days, snowflake automatically forces users to change their password upon next login, which will help us enforce the policy

PASSWORD_LIFETIME = 60

Preventing Reuse: Disallowing immediate reuse of old passwords is a best practice. Snowflake allows us to configure how many previous passwords are remembered to prevent reuse. Our plan is to restrict reusing last 3 passwords

PASSWORD_HISTORY_NUM_TO_CHECK = 3

3. Account Lockout:

Maximum Retries: Limiting the number of failed login attempts before locking the account is essential. Snowflake supports a range of 1 to 10 attempts. Our plan is to limit to 5 attempts.

PASSWORD_MAX_RETRIES = 5

Lockout Duration: Setting a reasonable lockout duration to deter brute-force attacks is a best practice

PASSWORD_LOCKOUT_TIME = 30 (30 minutes)

4. Break-Glass Accounts:

Our plan to is use the same password policy for the normal and break-glass accounts.
But it is possible to setup a dedicated policy which is more stringent. Also limiting the accounts only to have the absolutely necessary privileges is critical.

5. Additional rules : Our goal is to always enable MFA for all users, including break-glass accounts, for an extra layer of security when the password polices are fully enforced / in place. Also using password manager to generate strong passwords is going to be recommended to all users. Regularly audit password policies (yearly ?) and user activity to identify potential security gaps is also a consideration. But this will be taken up in the near future.

Related SQL Script to apply the above rules - 

-- Create a password policy for regular users

CREATE OR REPLACE PASSWORD POLICY user_password_policy
  PASSWORD_MIN_LENGTH = 14
  PASSWORD_REQUIRE_UPPERCASE = true
  PASSWORD_REQUIRE_LOWERCASE = true
  PASSWORD_REQUIRE_NUMERIC = true
  PASSWORD_REQUIRE_SPECIAL_CHAR = true
  PASSWORD_LIFETIME = 60
  PASSWORD_HISTORY_NUM_TO_CHECK = 5
  PASSWORD_MAX_RETRIES = 5
  PASSWORD_LOCKOUT_TIME = 30;

-- Alter users to use the appropriate policy 
ALTER USER <user> SET PASSWORD_POLICY = user_password_policy 

**I plan to do this recursively using a loop for all users**

-- Grant appropriate roles the ability to use the policies
GRANT USAGE ON PASSWORD POLICY user_password_policy TO ROLE sysadmin; 
GRANT USAGE ON PASSWORD POLICY user_password_policy TO ROLE securityadmin; 

-- Set the default password policy for new users
ALTER ACCOUNT SET DEFAULT_PASSWORD_POLICY = user_password_policy;

Also here a few important considerations, the next steps are heavily dependent on these -

- Apply this policy and enforcing this policy is efficient when SCIM integration is not in place. So we may be able to use this agency accounts in most cases
- When we have Okta integrated with Snowflake, things get a bit more complex
- In this case, Okta's password complexity requirements need to align with Snowflake password policies
- When using SCIM, 

1. Carefully coordination of password policies between Okta and Snowflake is critical to providing a seamless login experience for users
2. There might be a need to synchronize password changes between Okta and Snowflake. This involves configuring Okta to update Snowflake user passwords whenever they are changed in Okta
3. Also need to ensure the DEFAULT_PASSWORD_POLICY for the Snowflake account is set to the appropriate policy during automated provisioning 
4. We may also need to configure Okta to send the necessary attributes to Snowflake during provisioning to assign the correct password policy to the user.

**Can you please let me know your feedback on this policy ?**

**Comment by ian-r-rose** [comment]
Looking great! Here is some specific feedback:

> Following are best practices for password rules in Snowflake, based on their official documentation, to enforce strong security for regular and break-glass accounts -
> 
>     1. Password Complexity:
> 
> 
> Length: Enforcing a minimum password length is critical. Snowflake supports a range of 8 to 256 characters. A minimum of **14** characters is recommended. PASSWORD_MIN_LENGTH = 14
> 
> Character Diversity: Our policy wille a mix of uppercase, lowercase, numbers, and special characters (!@#$%^&* etc.). PASSWORD_REQUIRE_UPPERCASE = true PASSWORD_REQUIRE_LOWERCASE = true PASSWORD_REQUIRE_NUMERIC = true PASSWORD_REQUIRE_SPECIAL_CHAR = true

This all seems good to me. I suggest we add this to the terraform configuration `main.tf`, for the production deployments, what do you think @ram-kishore-odi? I don't think it makes sense to do account-level things in the 'etl' module, since we deploy it (at least) twice in the account, but it could be at the top-level.

> 
>     2. Password Expiration:
> 
> 
> Regular Expiration: Setting an appropriate password expiration interval is also very important. Snowflake's default is 90 days. Our plan is consider using a shorter interval like 60 for enhanced security. If the passwords are older than 60 days, snowflake automatically forces users to change their password upon next login, which will help us enforce the policy
> 
> PASSWORD_LIFETIME = 60

I was under the impression that expiring passwords were no longer considered best-practice for most contexts. Do Snowflake or IT-Ops have any guidance here?

> Our plan to is use the same password policy for the normal and break-glass accounts. But it is possible to setup a dedicated policy which is more stringent. Also limiting the accounts only to have the absolutely necessary privileges is critical.
> 

I think it's okay to use the same policy for both, the above plan looks plenty secure to me.

>     5. Additional rules : Our goal is to always enable MFA for all users, including break-glass accounts, for an extra layer of security when the password polices are fully enforced / in place. Also using password manager to generate strong passwords is going to be recommended to all users. Regularly audit password policies (yearly ?) and user activity to identify potential security gaps is also a consideration. But this will be taken up in the near future.
> 

Sounds good!

> 
> Related SQL Script to apply the above rules -
> 
> -- Create a password policy for regular users
> 
> CREATE OR REPLACE PASSWORD POLICY user_password_policy PASSWORD_MIN_LENGTH = 14 PASSWORD_REQUIRE_UPPERCASE = true PASSWORD_REQUIRE_LOWERCASE = true PASSWORD_REQUIRE_NUMERIC = true PASSWORD_REQUIRE_SPECIAL_CHAR = true PASSWORD_LIFETIME = 60 PASSWORD_HISTORY_NUM_TO_CHECK = 5 PASSWORD_MAX_RETRIES = 5 PASSWORD_LOCKOUT_TIME = 30;
> 
> -- Alter users to use the appropriate policy ALTER USER SET PASSWORD_POLICY = user_password_policy
> 
> **I plan to do this recursively using a loop for all users**

Is it possible to apply the default password policy to all users at once rather than altering each one individually? This looks annoying to maintain.

> -- Grant appropriate roles the ability to use the policies GRANT USAGE ON PASSWORD POLICY user_password_policy TO ROLE sysadmin; GRANT USAGE ON PASSWORD POLICY user_password_policy TO ROLE securityadmin;
> 
> -- Set the default password policy for new users ALTER ACCOUNT SET DEFAULT_PASSWORD_POLICY = user_password_policy;

Do we need to grant usage on the password policy to `sysadmin`? I am not sure that role needs access from the perspective of least-privilege.

I suggest we also look into putting these into the terraform configuration.

> 
> Also here a few important considerations, the next steps are heavily dependent on these -
> 
>     * Apply this policy and enforcing this policy is efficient when SCIM integration is not in place. So we may be able to use this agency accounts in most cases

Yeah, this isn't really relevant to Okta-managed accounts, it's mostly for users in other departments and in break-glass situations.

> 
>     * When we have Okta integrated with Snowflake, things get a bit more complex
> 
>     * In this case, Okta's password complexity requirements need to align with Snowflake password policies

I don't think this is true. IT-Ops has its own password complexity requirements for Okta, and we can just defer to that without worrying about the Snowflake account policy.

>     * When using SCIM,
> 
> 
>     1. Carefully coordination of password policies between Okta and Snowflake is critical to providing a seamless login experience for users

I also don't think this is true. The only people who will be able to use Okta or local login are admins in break-glass situations. So long as both password policies are reasonable, I don't think it's necessary to carefully coordinate them.

> 
>     2. There might be a need to synchronize password changes between Okta and Snowflake. This involves configuring Okta to update Snowflake user passwords whenever they are changed in Okta

I don't think this is necessary: SCIM *can* push passwords to user accounts, but I'm not sure it has to, and in fact might be undesirable (reuse of passwords and all that). As long as break-glass users are using a password manager, I see no reason to enforce Okta and Snowflake passwords to be the same.

**Comment by ram-kishore-odi** [comment]
> Looking great! Here is some specific feedback:
> 
> > Following are best practices for password rules in Snowflake, based on their official documentation, to enforce strong security for regular and break-glass accounts -
> > ```
> > 1. Password Complexity:
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Length: Enforcing a minimum password length is critical. Snowflake supports a range of 8 to 256 characters. A minimum of **14** characters is recommended. PASSWORD_MIN_LENGTH = 14
> > Character Diversity: Our policy wille a mix of uppercase, lowercase, numbers, and special characters (!@#$%^&* etc.). PASSWORD_REQUIRE_UPPERCASE = true PASSWORD_REQUIRE_LOWERCASE = true PASSWORD_REQUIRE_NUMERIC = true PASSWORD_REQUIRE_SPECIAL_CHAR = true
> 
> This all seems good to me. I suggest we add this to the terraform configuration `main.tf`, for the production deployments, what do you think @ram-kishore-odi? I don't think it makes sense to do account-level things in the 'etl' module, since we deploy it (at least) twice in the account, but it could be at the top-level.
> 
Hi @ian-r-rose, Yes, I agree. It's generally considered better to manage Snowflake password policies in Terraform scripts rather than SQL scripts. Your approach sounds good to me. I will open a new issue for this.

> > ```
> > 2. Password Expiration:
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Regular Expiration: Setting an appropriate password expiration interval is also very important. Snowflake's default is 90 days. Our plan is consider using a shorter interval like 60 for enhanced security. If the passwords are older than 60 days, snowflake automatically forces users to change their password upon next login, which will help us enforce the policy
> > PASSWORD_LIFETIME = 60
> 
> I was under the impression that expiring passwords were no longer considered best-practice for most contexts. Do Snowflake or IT-Ops have any guidance here?
> 
While Snowflake allows you to set password expiration policies, they don't explicitly recommend a specific expiration period. Their documentation focuses on password complexity requirements and other security measures.

You are correct. The consensus among security researchers and organizations like NIST has shifted away from recommending frequent password expirations.

I think instead of relying solely on password expiration, focussing on a holistic approach to password security that emphasizes strong passwords, MFA, and educating users appropriately is a going to work better overall

> > Our plan to is use the same password policy for the normal and break-glass accounts. But it is possible to setup a dedicated policy which is more stringent. Also limiting the accounts only to have the absolutely necessary privileges is critical.
> 
> I think it's okay to use the same policy for both, the above plan looks plenty secure to me.
> 
> > ```
> > 5. Additional rules : Our goal is to always enable MFA for all users, including break-glass accounts, for an extra layer of security when the password polices are fully enforced / in place. Also using password manager to generate strong passwords is going to be recommended to all users. Regularly audit password policies (yearly ?) and user activity to identify potential security gaps is also a consideration. But this will be taken up in the near future.
> > ```
> 
> Sounds good!
> 
> > Related SQL Script to apply the above rules -
> > -- Create a password policy for regular users
> > CREATE OR REPLACE PASSWORD POLICY user_password_policy PASSWORD_MIN_LENGTH = 14 PASSWORD_REQUIRE_UPPERCASE = true PASSWORD_REQUIRE_LOWERCASE = true PASSWORD_REQUIRE_NUMERIC = true PASSWORD_REQUIRE_SPECIAL_CHAR = true PASSWORD_LIFETIME = 60 PASSWORD_HISTORY_NUM_TO_CHECK = 5 PASSWORD_MAX_RETRIES = 5 PASSWORD_LOCKOUT_TIME = 30;
> > -- Alter users to use the appropriate policy ALTER USER SET PASSWORD_POLICY = user_password_policy
> > **I plan to do this recursively using a loop for all users**
> 
> Is it possible to apply the default password policy to all users at once rather than altering each one individually? This looks annoying to maintain.
> 
Yes, I agree. As you can see in my comment above, I indicated that I was going to do it recursively. Just did not include the script here.
> > -- Grant appropriate roles the ability to use the policies GRANT USAGE ON PASSWORD POLICY user_password_policy TO ROLE sysadmin; GRANT USAGE ON PASSWORD POLICY user_password_policy TO ROLE securityadmin;
> > -- Set the default password policy for new users ALTER ACCOUNT SET DEFAULT_PASSWORD_POLICY = user_password_policy;
> 
> Do we need to grant usage on the password policy to `sysadmin`? I am not sure that role needs access from the perspective of least-privilege.
> 
> I suggest we also look into putting these into the terraform configuration.
> 
Yes I agree, having all this stuff in terraform is good ! We can have one story for the password policy implementation including the core policy and implementation

The sysadmin and securityadmin roles are typically responsible for maintaining security standards, so I felt it makes it makes sense for them to have the ability to manage password policies.

> > Also here a few important considerations, the next steps are heavily dependent on these -
> > ```
> > * Apply this policy and enforcing this policy is efficient when SCIM integration is not in place. So we may be able to use this agency accounts in most cases
> > ```
> 
> Yeah, this isn't really relevant to Okta-managed accounts, it's mostly for users in other departments and in break-glass situations.
> 
> > ```
> > * When we have Okta integrated with Snowflake, things get a bit more complex
> > 
> > * In this case, Okta's password complexity requirements need to align with Snowflake password policies
> > ```
> 
> I don't think this is true. IT-Ops has its own password complexity requirements for Okta, and we can just defer to that without worrying about the Snowflake account policy.
> 

Yes correct, if we are going to stick to IT-Ops password complexity requirements for Okta, then we can defer to that.
Since we are working towards Okta SCIM integration with Snowflake, this is even more applicable to our scenario.

> > ```
> > * When using SCIM,
> > 
> > 
> > 1. Carefully coordination of password policies between Okta and Snowflake is critical to providing a seamless login experience for users
> > ```
> 
> I also don't think this is true. The only people who will be able to use Okta or local login are admins in break-glass situations. So long as both password policies are reasonable, I don't think it's necessary to carefully coordinate them.
> 

Yes, once Okta SCIM integration with Snowflake is in place, user password can be pushed from Okta into Snowflake, **if required**. So in this case no coordination is needed 

> > ```
> > 2. There might be a need to synchronize password changes between Okta and Snowflake. This involves configuring Okta to update Snowflake user passwords whenever they are changed in Okta
> > ```
> 
> I don't think this is necessary: SCIM _can_ push passwords to user accounts, but I'm not sure it has to, and in fact might be undesirable (reuse of passwords and all that). As long as break-glass users are using a password manager, I see no reason to enforce Okta and Snowflake passwords to be the same.

Yes correct, and especially given the break-glass users will use a password manager (above comment is also applicable here)

**Comment by ian-r-rose** [comment]
> Yes, I agree. As you can see in my comment above, I indicated that I was going to do it recursively. Just did not include the script here.

I don't think I understand what you mean by recursively. I think of a function that calls itself.

> The sysadmin and securityadmin roles are typically responsible for maintaining security standards, so I felt it makes it makes sense for them to have the ability to manage password policies.

I agree that securityadmin is responsible for security standards, but I don't agree that sysadmin is. My understanding of the [system-defined roles](https://docs.snowflake.com/en/user-guide/security-access-control-overview#system-defined-roles) is that sysadmin is mostly for creating databases and warehouses, and not intended for security operations.

**Comment by ram-kishore-odi** [comment]
> > Yes, I agree. As you can see in my comment above, I indicated that I was going to do it recursively. Just did not include the script here.
> 
> I don't think I understand what you mean by recursively. I think of a function that calls itself.
> 
I am sorry that was not the right word to use. I meant doing it for all users at once instead of running alter statements individually. I think we are on the same page here.

> > The sysadmin and securityadmin roles are typically responsible for maintaining security standards, so I felt it makes it makes sense for them to have the ability to manage password policies.
> 
> I agree that securityadmin is responsible for security standards, but I don't agree that sysadmin is. My understanding of the [system-defined roles](https://docs.snowflake.com/en/user-guide/security-access-control-overview#system-defined-roles) is that sysadmin is mostly for creating databases and warehouses, and not intended for security operations.

Looking at the system-defined roles, I now agree that sysadmin would not need the usage grant on the password policy.
It is more appropriate for the securityadmin to have it. Thank you!! I will make sure that the terraform config only grants it to securityadmin.

**Comment by ian-r-rose** [comment]
Sounds good! Closing as complete, implementation will be tracked in #455
